Tab redesign

jared-hughes opened the pull request:
This is an implementation of #655 (Tab redesign). A few minor rough edges (described below), but this PR is mergeable.

I overwrote the existing ng-hole code to make use of the existing separation between hole and ng-hole.

## Sample screenshots

The default layout.

![image](https://user-images.githubusercontent.com/20214911/169651691-92dd1484-c5e2-4fb9-95fe-770116789ea3.png)

The "add row" button can be used to add the first item of the pool to a new row. This is mainly for mobile accessibility, but it's also a convenience on desktop. (I didn't log in here; I just hid the alert with the close button.)

![image](https://user-images.githubusercontent.com/20214911/169651794-fe3922ff-7670-4b0a-998f-0da499486a92.png)

Layout after some customization.

The layout can be reorganized by dragging tabs or using the click interactions: close, maximize, and add back from the tab pool. Dragging the splitter between two stacks in a row/column can change their relative sizes. In addition, the "wide" layout button causes the layout to take the full width.

![image](https://user-images.githubusercontent.com/20214911/169651728-acd540eb-9dc1-4eb0-bae2-f15202b29d68.png)

Appearance after submitting:

![image](https://user-images.githubusercontent.com/20214911/169651774-32faa739-50d6-44ee-9489-fde3acf5b7e7.png)

Demonstration of dragging a tab in from the pool
![image](https://user-images.githubusercontent.com/20214911/168062640-30e59c5a-408d-4e45-a9c5-6db9314b32c5.png)

Any tab can be maximized
![image](https://user-images.githubusercontent.com/20214911/169651884-700a8ca4-02f1-4659-bec4-46cfbea5d44d.png)

Dark theme is supported
![image](https://user-images.githubusercontent.com/20214911/169651925-004ae839-bdcb-4315-84d3-0338686c64b9.png)

## How merging affects the rest of the site

Only the ng hole path is significantly visually changed; everything else is almost the same. The only difference is the diff view on normal hole editor: the headers are black on white (in light mode) rather than large and white on black. Still looks good.

The blank div in the header now has id `alert-pool` to allow for placing alerts.

In dark mode, scrollbars across the site are now dark (this doesn't affect Firefox, which seems to have detected scrollbar color already). This is included in the PR because the tab layout can end up with a bunch of smaller scrollbars, so the light-color scrollbars are very out of place.

The mini-scoreboard code (`apiMiniRankingsGET`) in `api.go` was modified to increase the limit from 7 to 99 on ng. I also rewrote the me-centering code to avoid the O(N^2) for-loop. Should have the same behavior.

## Improvements to-do

- [ ] Improve default layout as discussed in #655. Maybe also close output panels by default, then show them once the code has been run, if all are still closed
- [ ] Layout persistence (layout is not currently saved between page loads)
- [x] Ability to return to default view (maybe button in header)
  - [x] ~~This could also allow resetting the code to the sample code~~ Not related to layout, abandoning for now
- [x] Details needs some padding
- [x] Code size and "ctrl+enter or Run" should be moved to the bottom of the code panel
  - <s>maybe move it inside the scrollbar as well, to get more vertical height. Unfortunately this obscures the byte count at some scroll positions</s>
- [x] Scoreboard isn't currently responsive to resizing
- [x] ~~Drag tabs to the tab pool? Need `constrainDragToContainer=false`~~ No need. Just press the close button <kbd>×</kbd>
- [x] Let horizontal full-width be toggleable
- [x] Smarter tab insertion when clicking a tab in the tab pool. Update: took a simple direction that turned out good
- [x] Allow collapsing language selector
- [x] Bug: dragging from pool away from the layout leads to a lm_dragProxy that doesn't get cleared; sometimes throws "Ground node can only have a single child"

## Vertical Height

A major issue faced with the tab layout is vertical height. If the main page is allowed to scroll, drag interactions become finicky, and the tab layout is hard to define. I went with Golden Layout's direction and disabled scrolling on the main page. Unfortunately this means not much space is available for the full layout, considering there's many headers.

A few improvements help mitigate the vertical height problem:
- maximizing tabs allows for any to be looked at individually at full size
- hiding alerts helps for experimental holes and people not logged-in
- some vertical grid-gaps are reduced
- The solution selector, tab pool, and pass/fail status display are placed next to each other
- The language selector can be collapsed

However, this isn't enough for small screens if the user wants more than 1 row on the layout. For example, on my 1366x768 laptop with default chrome tabs and with bookmarks hidden, only 6.5 lines are available to the code in the current default layout. This is half of the 13 lines of the current production layout.

![image](https://user-images.githubusercontent.com/20214911/168066406-275d1740-039f-4b03-bb78-9729ef809349.png)

The height is acceptable when maximized:

![image](https://user-images.githubusercontent.com/20214911/168066870-a25553bf-2157-4b5e-8b77-e0efed2a2c14.png)

## Mobile

This is mobile-accessible. The layout automatically re-flows to vertical (rows become columns), more scrolling is allowed, and dragging tabs is disallowed (because it is hard to allow at the same time as scrolling the main page)

![image](https://user-images.githubusercontent.com/20214911/169652003-7afe47a4-bd63-4b55-8bfc-a7d5326a9e8f.png)

- [x] make the default layout vertical-only: everything in one column
- [x] allow scrolling on the full page
- [x] disallow dragging tabs or row/column splitters
  - [ ] Considering disabling dragging the splitters. Easy to accidentally drag the wrong thing, but there should be some way to resize.
- [x] Cater tab maximisation to the small screen

JRaspass commented:
Just a heads up, I may have screwed you a little by removing NG. I didn't like how much duplication the old experiment led to so I'd like to do things differently for this experiment, at the very least I don't want to duplicate the route/view. I was thinking instead we'd pull in your alternative JS/CSS based on a user setting. Anyways let's discuss this in Discord in the coming weeks, going to give people a little time to get used to CM6 in case we need to revert or something.

jared-hughes commented:
Now the tab layout can be chosen as a user setting (I copied the code from the old keymap option). Internally, `render.go` switches to the `hole-tabs` path from `hole` if the user enabled tab layout in settings.

Also, to avoid duplication, I removed the css that `hole-tabs.css` had duplicated from `hole.css` in favor of prepending `hole.css` (fbcd0c8) and moved the common js code of `hole.tsx` and `hole-tabs.tsx` to `_hole-common.tsx` (a532fed). The js code sharing is a bit confusing because it requires passing around various callbacks for editing HTML, but at least the main logic (submissions, local storage solutions, byte counts, etc.) is de-duplicated well.

jared-hughes commented:
The e2e is passing now, so this should be mergeable after review.

JRaspass commented on css/dark.css:
```diff
@@ -1,46 +1,52 @@
 }
 
 body > header {
-    --background: #ddd;
```
why has the indent on this file changed to two?
